datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -584,23 +544,7 @@ fn get_valid_types(
                 match target_type_class {
                     TypeSignatureClass::Native(native_type) => {
                         let target_type = native_type.native();
-                        if &logical_type == target_type {
```

Comment:
Does that mean others function that used Coercible String now also cast integer to string?

Reply:
`TypeSignature::Coercible` is designed to cast between the same logical type, but now it cast any kind of type, I don't think this is ideal.

Reply:
For example, in this PR, `SELECT bit_length(12);` now return 16 instead of error, but I think we should error. Any unexpected type is valid which doesn't seem correct

Reply:
@shehabgamin 
I suggest with revert to the previous change and add binary->string casting given most of the string function allow this kind of casting, but not int->string like `ascii(2)`

For Signature::String, we can make it a convenient wrapper of `Coercible(string, string)` or deprecate it.

Reply:
The rule of thumb is that try to be consistent with either Postgres and DuckDB for the functions and don't make the casting too general that cast any type to any type since this turns invalid query into valid one and it is not easy to know that given the test is not covered.

Based on that,
1. Cast from Binary to String 👍🏻 
2. Cast from Numeric to String 👎🏻  (At least for the string functions changed in this PR)
3. Cast from Float to Integer 👎🏻 (At least not for `repeat`)